### PR TITLE
introducing vertical rhythm

### LINKED ISSUE #550: Layout is a pain
Unfortunately we have inconsistent dimension, sizes and distances. Some of them are caused by evil `collapsing-margins`.

I created this issue to collect all the needed use-cases and to propose a solution, which will be properly defined and documented in the style-guide.
## Feature Request:
- [ ] Cross-Browser friendly
- [ ] Responsive
- [ ] No unexpected distances (eliminate collapsing margins)
- [ ] Lego principle - the Goal is that we end up with simple blocks which can be stacked or aligned in a row
- [ ] Markup generated by WYSIWYG-Editor should be considered
## Collapsing Margins:

Those are one of the main cause of our style issues, like:
- layout - e.g. distance to other nodes
- dimension - e.g. parent eats childs margin and thus does not grow
- backgrounds - e.g. hence parents eats margin, backgrounds get clipped

You can find a detailed explanation of why they happen here https://css-tricks.com/what-you-should-know-about-collapsing-margins/
## Proposed Solution:
- Margins go always only in one direction (up or down)
- Each Component has a wrapping element which defines it's distance to others through `padding` (cause paddings do not collapse)

In BEM words:
- Blocks do not use `margin`, just `padding`, `border`, etc.
- Only elements do use `margin` too, but only in one direction

**Basic Component**
component

**List**
list

**List-2-Column**
list-2-column

**List-3-Column**
list-3-column

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #550

### CODE REVIEW

**Comment by sventschui** [comment]
do you plan to merge this like it is or are more features planned prior to merging this?

**Comment by AndyOGo** [comment]
@sventschui 
hehe, it's WIP.
I'm also arranging some details with @anja225

**Comment by AndyOGo** [comment]
@davidknezic 
@sventschui 
Why do `.button`'s have `10px` top and bottom margin?
I would default it to `0`

**Comment by sventschui** [comment]
Because that will break a lot of stuff when buttons wrap to a new line. They need a margin in at least one direction

**Comment by AndyOGo** [comment]
@sventschui 
I see. Can't give the parent some margin or padding?
`margin-top` would be actually. actually at least the badges need a top margin...

**Comment by AndyOGo** [comment]
Now I have still 4 blocks which utilize `margin-bottom`, though I can't find demo pages which implement those blocks.

- [ ] [button-row.scss](https://github.com/axa-ch/style-guide/blob/develop/scss/style/blocks/button-row.scss#L10)
- [ ] [fieldset.scss](https://github.com/axa-ch/style-guide/blob/develop/scss/style/blocks/fieldset.scss#L9)
- [ ] file-upload-single.scss
- [ ] heading.scss

**Comment by AndyOGo** [comment]
Okay, there is a [`fieldset.pug`](https://github.com/axa-ch/style-guide/blob/develop/docs/components/form-and-input-elements/fieldset.pug), but it's empty...

**Comment by sventschui** [comment]
> I see. Can't we give the parent some margin or padding?

The case is the following: 

```
div
  .button
  .button
```

If these wrap, they need a margin-top or -bottom

**Comment by AndyOGo** [comment]
@sventschui 
Thank you very much. then I will go for `margin-top` 😄

**Comment by AndyOGo** [comment]
@sventschui @davidknezic @elessar-ch 
I investigate more on my own whats up with the 4 remaining blocks, here are my findings and suggested actions:

# [button-row.scss](https://github.com/axa-ch/style-guide/blob/develop/scss/style/blocks/button-row.scss#L10)
I couldn't find code which is actively using this block within the Style-Guide.

**BUT** it is used for AEM Forms in code like:
`/aem-all_ext/apps/forms/content/src/main/content/jcr_root/apps/axa/forms/components/controls/actions/nextitemnav/widget.jsp`
```jsp
<%@ page import="com.adobe.aemds.guide.utils.GuideConstants" %>
<%@include file="/libs/fd/af/components/guidesglobal.jsp" %>
<%GuideNode guideField =(GuideNode) request.getAttribute("guideField");%>

<div class="<%= GuideConstants.GUIDE_FIELD_WIDGET%> button button-row__button--right">
    ${guide:encodeForHtml(guideField.title,xssAPI)}
</div>
```
## Suggested Action
- [ ] Keep it and add missing example for illustration purpose.

# [fieldset.scss](https://github.com/axa-ch/style-guide/blob/develop/scss/style/blocks/fieldset.scss#L9)

I couldn't find code which is actively using this block within the Style-Guide nor AEM.
Also couldn't find any design specifications in confluence.

## Suggested Action

- [ ] Kill it 🔫 

# [file-upload-single.scss](https://github.com/axa-ch/style-guide/blob/develop/scss/style/blocks/file-upload-single.scss#L26)

This block is illustrated in `fileupload-single.pug` within the StyleGuide, though it's not used inside AEM.

## Suggested Action
- [ ] Clean it up and keep it.

# [heading.scss](https://github.com/axa-ch/style-guide/blob/develop/scss/style/blocks/heading.scss#L13-L83)

This block implements various modifiers for the `.heading` block.
I found the following details:

- `.heading--page`
  This modifier seems to be only used for the heading of pages within the styleguide generate with the `page.pug` layout. Right?
  - [ ] Keep it - maybe move it to style-guide related `scss`
- `.heading--modal`
  IMHO block related headings should be styled with that block and not another one, which actually exists `modal__heading`, but it is implemented in a strange way:
  - [ ] Fix it 🔫 
- `.heading--fieldset`
  I couldn't find any usage neither within the StyleGuide nor on AEM.
  - [ ] Kill it 🔫 
- `.heading--summary`
  I couldn't find any usage neither within the StyleGuide nor on AEM.
  - [ ] Kill it 🔫

**Comment by AndyOGo** [comment]
@sventschui @davidknezic @elessar-ch 
Your Feedback is very very welcome and needed, please 🙏

**Comment by AndyOGo** [comment]
I just found moar unused blocks...

## Killlist

- [ ] `.example-header`

File: docs/components/snippets/icon-teaser.pug [hunk 1]
```diff
@@ -16,5 +16,5 @@ include ../../mixins.pug
         .icon-teaser__item__icon-wrapper
           +svg(item.icon, ['icon-teaser__item__icon'])
         .icon-teaser__item__title= item.title
```
Comment (LucaMele) [inline docs/components/snippets/icon-teaser.pug]: why does item__text get a heading element and item__title not?

File: scss/style/blocks/form.scss [hunk 2]
```diff
@@ -102,6 +97,27 @@
...
 
+.form__group--condensed {
+  margin-top: 14px;
+
+  &:first-child {
+    margin-top: 0;
+  }
+
+  .form__group__label {
+    font-size: 0.9285em;
```
Comment (LucaMele) [inline scss/style/blocks/form.scss]: em?

File: scss/style/blocks/form.scss [hunk 3]
```diff
@@ -102,6 +97,27 @@
...
+    color: $color-gray--darker;
+
+    @include respond(medium) {
+      font-size: 1em;
```
Comment (LucaMele) [inline scss/style/blocks/form.scss]: em?

Comment (AndyOGo) [inline scss/style/blocks/form.scss]: Mhm, I can only guess why. Maybe to change font-size relatively by intention.

**Pull request merged**